clients/kafka-common/src/main/java/org/eclipse/hono/kafka/client/CachingKafkaProducerFactory.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+package org.eclipse.hono.kafka.client;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+import org.apache.kafka.common.errors.AuthorizationException;
+import org.apache.kafka.common.errors.OutOfOrderSequenceException;
+import org.apache.kafka.common.errors.ProducerFencedException;
+import org.apache.kafka.common.errors.UnsupportedForMessageFormatException;
+import org.apache.kafka.common.errors.UnsupportedVersionException;
+import org.eclipse.hono.kafka.client.test.FakeProducer;
+
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import io.vertx.kafka.client.producer.KafkaProducer;
+
+/**
+ * A factory for creating Kafka producers.
+ * <p>
+ * This implementation provides no synchronization and should not be used by multiple threads.
+ * <p>
+ * Created producers are being cached.
+ * <p>
+ * Producers are closed and removed from the cache if they throw a {@link #isFatalError(Throwable) fatal exception}. A
+ * following invocation of {@link #getOrCreateProducer(String, Map)} will then return a new instance.
+ *
+ * @param <K> The type for the record key serialization.
+ * @param <V> The type for the record value serialization.
+ */
+public class CachingKafkaProducerFactory<K, V> {
```

Comment:
AFAIK `CachingClientFactory` is quite generic and I think it can also accommodate Kafka clients and in this case `KafkaProducer`? Then we could reuse the code and IMHO having a uniform style throughout, makes the code easier to understand. I am just curious to know if there are any reasons to have a new  `CachingKafkaProducerFactory`.

Reply:
CachingClientFactory was not public when I wrote the code.
FMPOV the `CachingKafkaProducerFactory` is not comparable to `CachingClientFactory` but more to the sender factories like `org.eclipse.hono.client.impl.DownstreamSenderFactoryImpl`. It might be that the naming causes confusion.
I could have used `CachingClientFactory` inside of `CachingKafkaProducerFactory` to hold the producer instances. But it does not fit well because it is designed around managing a connection and synchronizing expensive client creation. Using a shared KafkaProducer all of this is already provided by `io.vertx.kafka.client.producer.impl.KafkaProducerImpl`, so I think a simple map is sufficient to hold the producer.